rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp

```diff
@@ -402,30 +420,61 @@ void SqliteStorage::prepare_for_reading()
         topic_list += ",";
       }
     }
-    statement_str += "(topics.name IN (" + topic_list + ")) AND ";
+    where_conditions.push_back("(topics.name IN (" + topic_list + "))");
   }
   // add topic filter based on regular expression
   if (!storage_filter_.topics_regex.empty()) {
     // Construct string for selected topics
-    statement_str += "(topics.name REGEXP '" + storage_filter_.topics_regex + "')";
-    statement_str += " AND ";
+    where_conditions.push_back("(topics.name REGEXP '" + storage_filter_.topics_regex + "')");
   }
   // exclude topics based on regular expressions
   if (!storage_filter_.topics_regex_to_exclude.empty()) {
     // Construct string for selected topics
-    statement_str += " (topics.name NOT IN ";
-    statement_str += " (SELECT topics.name FROM topics WHERE topics.name REGEXP '";
-    statement_str += storage_filter_.topics_regex_to_exclude + "')";
-    statement_str += " ) AND ";
+    where_conditions.push_back(
+      "(topics.name NOT IN "
+      "(SELECT topics.name FROM topics WHERE topics.name REGEXP '" +
+      storage_filter_.topics_regex_to_exclude + "'))");
+  }
+
+  const std::string direction_op = read_order_.reverse ? "<" : ">";
+  const std::string order_direction = read_order_.reverse ? "DESC" : "ASC";
+
+  // add seek head filter
+  if (read_order_.sort_by == rosbag2_storage::ReadOrder::File) {
+    where_conditions.push_back(
+      "(messages.id " + direction_op + "= " + std::to_string(seek_row_id_) + ") ");
+  } else {
+    where_conditions.push_back(
+      "(((timestamp = " + std::to_string(seek_time_) + ") "
+      "AND (messages.id " + direction_op + "= " + std::to_string(seek_row_id_) + ")) "
+      "OR (timestamp " + direction_op + " " + std::to_string(seek_time_) + ")) ");
+  }
+
+  for (
+    std::vector<std::string>::const_iterator it = where_conditions.begin();
+    it != where_conditions.end(); ++it)
+  {
+    statement_str += *it;
+    if (it != where_conditions.end() - 1) {
+      statement_str += " AND ";
+    }
   }
-  // add start time filter
-  statement_str += "(((timestamp = " + std::to_string(seek_time_) + ") "
-    "AND (messages.id >= " + std::to_string(seek_row_id_) + ")) "
-    "OR (timestamp > " + std::to_string(seek_time_) + ")) ";
 
   // add order by time then id
-  statement_str += "ORDER BY messages.timestamp, messages.id;";
+  switch (read_order_.sort_by) {
+    case rosbag2_storage::ReadOrder::ReceivedTimestamp:
+      statement_str += "ORDER BY messages.timestamp " + order_direction;
+      statement_str += ", messages.id " + order_direction;
```

Comment:
The same as previous comment
```suggestion
      statement_str += "ORDER BY messages.timestamp " + order_direction;
```

Reply:
Same as previous response - we want to ensure that reverse timestamp order is exactly reverse to the forward direction:


```
1: 100
2: 100
3: 200
```

should return `3, 2, 1` because the forward direction returns `1, 2, 3`. But, the base index of the database is by rowid, so with this suggestion, it would probably return `3, 1, 2`

Reply:
I do notice now that there isn't a test for this, though. So, I should add one to make sure of it.

Reply:
Agree with your explanation about ordering. Unit tests would be very appreciated.

Reply:
Test added for this case!